### PR TITLE
immutable while condition

### PR DESCRIPTION
**Pull request by kimsnj**
This PR aims at implementing lint suggested in [#40(comment)](https://github.com/rust-lang-nursery/rust-clippy/issues/40#issuecomment-251462183).

Current status, lints when:
- [x] All variables in the body are immutable
- [x] Variables in the condition do not appear in the body
- [x] Variables only used by const ref or by value. Special case to handle:
     - [x] Shadowing
     - ~~[ ] Inner mutability (cell, refcell, mutex, sync::atomic, anything else?)~~

Is there any additional step or edge case to be handled?

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -2112,3 +2139,101 @@ fn path_name(e: &Expr) -> Option<Name> {
...
+}
+
+/// checks within an expression/statement if any of the variables are used mutably
+struct MutablyUsedVisitor<'a, 'tcx: 'a> {
+    cx: &'a LateContext<'a, 'tcx>,
+    ids: HashSet<NodeId>,
+    any_used: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for MutablyUsedVisitor<'a, 'tcx> {
```
Comment (kimsnj) [inline clippy_lints/src/loops.rs]: I just noticed there is [ExprUseVisitor](https://michael-f-bryan.github.io/rustc-internal-docs/rustc/middle/expr_use_visitor/struct.ExprUseVisitor.html).
Would it be safer to use or is the match on AddrOf and Assign expressions enough?
I am actually pretty confused by all the arguments required by ExprUseVisitor as well as what Delegate is suppose to do in consume/borrow methods…

**Comment by kimsnj** [comment]
I managed to make the lint work with `ExprUseVisitor`! \o/

However, given that I already bail out in case of function/method calls, I don't need to do any special handling for types with interior mutability. 
AFAIK, They do not implement any operator (returning `bool`) that can be directly used in the `while` condition. So we should be covered by the function calls exit.

**Comment by oli-obk** [comment]
> However, given that I already bail out in case of function/method calls, 

The tests suggest otherwise (functions taking immutable references are not treated as a mutation). In the case of interior mutability, this is not true.

Can you add some tests for interior mutability in any case?

**Comment by kimsnj** [comment]
I was referring to the step where I collect the variables used within the condition of the while loop.
As soon as we come across a function / method call we bail (https://github.com/rust-lang-nursery/rust-clippy/pull/2483/files#diff-808a9ae7b2b19869fb07162ff274d7dbR2197) as we have no way to know whether the function has a side-effect.

Given this, if we want to use an internally mutable type, we'll need to use a `get` or `load` method to access to the inner condition.
```rust
    let b = AtomicBool::new(true);
    while b {
    //    ^ expected bool, found struct `std::sync::atomic::AtomicBool`
   //       we have to use b.load(Ordering::Relaxed) in which case we stop before even checking the loop body
        b.store(false, Ordering::Relaxed);
    }
```

But, for sure, i will add a test to cover this case.
I'll do a PR this afternoon.

File: tests/ui/infinite_loop.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        i = 5;
+        println!("OK - i in cond and mentionned");
+    }
+}
+
+fn used_immutable() {
+    let mut i = 0;
+
+    while i < 3 {
+        fn_constref(&i);
```
Comment (oli-obk) [inline tests/ui/infinite_loop.rs]: I meant if this example were written with a `Cell`, the `fn_constref` function could modify it without us knowing about it.

File: tests/ui/infinite_loop.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+use std::cell::Cell;
+
+fn maybe_i_mutate(i: &Cell<bool>) { unimplemented!() }
+
+fn internally_mutable() {
+    let b = Cell::new(true);
+
+    while b.get() {       // b cannot be silently coerced to `bool`
```
Comment (oli-obk) [inline tests/ui/infinite_loop.rs]: ooooh. now I get what you were trying to say! Awesome!

**Comment by oli-obk** [comment]
This lint is absolutely golden!

**Comment by kimsnj** [comment]
Yay! Thanks!
I also really liked Boddlnagg's suggestion!

**Pull request merged**